pkg/applicationserver/io/pubsub/provider/connection.go

```diff
@@ -96,14 +85,41 @@ type Connection struct {
 
 // Shutdown shuts down the topics, subscriptions and the connections if required.
 func (c *Connection) Shutdown(ctx context.Context) error {
-	for _, s := range []Shutdowner{
+	return shutdown(ctx,
 		&c.Topics,
 		&c.Subscriptions,
 		c.ProviderConnection,
-	} {
+	)
+}
+
+var errShutdownFailed = errors.DefineInternal("shutdown_failed", "failed to shutdown")
+
+func shutdown(ctx context.Context, shutdowners ...Shutdowner) error {
+	details := make([]errors.ErrorDetails, 0, len(shutdowners))
+	for _, s := range shutdowners {
+		if isNil(s) {
+			continue
+		}
 		if err := s.Shutdown(ctx); err != nil {
-			return err
+			details = append(details, errShutdownFailed.WithCause(err))
```

Comment:
Now the error and its details are `errShutdownFailed`; I guess the details can just be `err` here